lib/response-handler.js

```diff
@@ -0,0 +1,168 @@
+"use strict";
+
+module.exports = (error, logger, parseLinkHeader) => {
+    /**
+     * @typedef {Object} artifactResponse
+     * @property {string} statusCode The HTTP status code.
+     * @property {string} message The status message for the response.
+     * @property {Object} body The response body.
+     * @property {string} location The url of the uploaded artifact.
+     */
+
+    /**
+     * Handles the response gotten from making requests to Shelf.
+     *
+     * @param {http~IncomingMessage} response
+     * @return {Promise.<Object>}
+     */
+    function handleArtifactResponse(response) {
+        var artifactResponse;
+
+        logger.debug("In handleArtifactResponse()");
+
+        if (!response.body) {
+            response.body = {};
+        }
+
+        artifactResponse = {};
+        logger.debug(`Response: ${JSON.stringify(response, null, 4)}`);
+        logger.debug(`Status code: ${response.statusCode}`);
+        logger.debug(`Response body: ${JSON.stringify(response.body, null, 4)}`);
+        artifactResponse.statusCode = response.statusCode;
+
+        if (artifactResponse.statusCode === 201) {
+            artifactResponse.message = "Artifact upload successful.";
+            artifactResponse.location = response.headers.location;
+        } else if (artifactResponse.statusCode === 200) {
+            artifactResponse.message = "Artifact download successful.";
+            artifactResponse.body = response.body;
+        } else {
+            if (response.body.code) {
+                // Try to grab the error code we received.
+                artifactResponse = Object.keys(error).filter((key) => {
+                    return key.code === response.body.code;
+                });
+
+                if (artifactResponse) {
+                    throw new Error(artifactResponse.message, artifactResponse.code);
+                } else {
+                    throw new Error(error.UNKNOWN.message, error.UNKNOWN.code);
+                }
+            } else if (response.body === error.UNAUTHORIZED.message) {
+                throw new Error(error.UNAUTHORIZED.message, error.UNAUTHORIZED.code);
+            }
+
+            throw new Error(error.UNKNOWN.message, error.UNKNOWN.code);
+        }
+
+        return Promise.resolve(artifactResponse);
```

Comment:
I need to resolve the promise if the user wants a `.then()` after calling a function on one of the shelflib objects (Artifact, Metadata, ArtifactSearch). I just tested it.